File: WdtcCore/src/main/java/org/wdt/wdtc/core/launch/LaunchProcess.java
package org.wdt.wdtc.core.launch;

import lombok.Setter;
import org.apache.log4j.Logger;
import org.wdt.utils.io.IOUtils;
import org.wdt.wdtc.core.download.infterface.TextInterface;
import org.wdt.wdtc.core.utils.ThreadUtils;
import org.wdt.wdtc.core.utils.WdtcLogger;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Setter
public class LaunchProcess {
  private static final Logger logmaker = WdtcLogger.getLogger(LaunchProcess.class);
  private final Process process;
  private TextInterface setUIText;

  public LaunchProcess(Process process) {
    this.process = process;
  }

  public void startLaunchGame() throws IOException {
    try {
      ThreadUtils.startThread(() -> getRunInfo(process.getInputStream()));
      ThreadUtils.startThread(() -> getRunInfo(process.getErrorStream())).join();
      logmaker.info("Game Stop");
    } catch (InterruptedException e) {
      logmaker.error("Run Command Error,", e);
    }

  }

  private void getRunInfo(InputStream inputStream) {
    try {
      BufferedReader Reader = new BufferedReader(new InputStreamReader(inputStream, "GBK"));
      String line;
      while ((line = Reader.readLine()) != null) {
        if (Thread.currentThread().isInterrupted()) {
          launchErrorTask();
          return;
        } else {
          Matcher ErrorWarn = Pattern.compile("FATAL").matcher(line);
          if (ErrorWarn.find()) {
            System.out.println(line);
            Thread.currentThread().interrupt();
          } else {
            System.out.println(line);
          }
        }
      }

    } catch (IOException e) {
      logmaker.error("Run Command Error,", e);
    }
  }

  private void launchErrorTask() throws IOException {
    setUIText.setControl("启动失败:\n" + IOUtils.toString(process.getErrorStream()) +
        IOUtils.toString(process.getInputStream()));
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/launch/LaunchProcess.kt
package org.wdt.wdtc.core.launch

import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.wdt.wdtc.core.utils.error
import org.wdt.wdtc.core.utils.logmaker
import java.io.IOException
import java.io.InputStream
import java.nio.charset.Charset
import java.util.regex.Pattern
import kotlin.system.measureTimeMillis

class LaunchProcess(
  private val process: ProcessBuilder,
  var printInfo: (String) -> Unit
) {
  private val builder = StringBuilder()
  fun startLaunchGame() {
    try {
      process.start().run {
        val runtime = measureTimeMillis {
          runBlocking(Dispatchers.IO) {
            launch(CoroutineName("Read info inputStream")) { getRunInfo(inputStream) }
            launch(CoroutineName("Read error inputStream")) { getRunInfo(errorStream) }
          }
        }
        logmaker.info("Game over. Game run time: $runtime ms")
      }
    } catch (e: InterruptedException) {
      logmaker.error("Run command error,", e)
    }
  }

  private fun getRunInfo(inputStream: InputStream) {
    try {
      val thread = Thread.currentThread()
      val reader = inputStream.reader(Charset.forName("GBK")).buffered()
      var line: String
      while (reader.readLine().also { line = it } != null) {
        line.let {
          if (thread.isInterrupted) {
            launchErrorTask()
            return
          } else {
            if (Pattern.compile("FATAL").matcher(it).find()) {
              println(it)
              builder.append(it)
              thread.interrupt()
            } else {
              println(it)
              builder.append(it)
            }
          }
        }
      }
    } catch (e: IOException) {
      logmaker.error("Run Command Error,", e)
    }
  }

  private fun launchErrorTask() {
    printInfo("启动失败:\n${builder}")
  }
}
